import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/AttributesUtilTest.java

```diff
@@ -0,0 +1,220 @@
+  @Test
+  public void testAttributesToQidoPath_simple() throws Exception {
+    Attributes attrs = new Attributes();
+    attrs.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+    attrs.setString(Tag.SOPInstanceUID, VR.UI, "123");
+
+    String result = AttributesUtil.attributesToQidoPath(attrs);
+
+    assertThat(result).isEqualTo(
+        "instances?" + TagUtils.toHexString(Tag.SOPInstanceUID) + "=123&");
+  }
+
+  @Test
+  public void testAttributesToQidoPath_urlEncode() throws Exception {
+    Attributes attrs = new Attributes();
+    attrs.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+    attrs.setString(Tag.PatientName, VR.CS, "%&#^ ");
+
+    String result = AttributesUtil.attributesToQidoPath(attrs);
+
+    assertThat(result).isEqualTo(
+        "instances?" + TagUtils.toHexString(Tag.PatientName) + "=%25%26%23%5E+&");
+  }
+
+  @Test
+  public void testAttributesToQidoPathArray_noModalitiesSet() throws Exception {
+    Attributes attrs = new Attributes();
+    attrs.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+
+    String[] results = AttributesUtil.attributesToQidoPathArray(attrs);
+
+    assertThat(results.length).isEqualTo(1);
+    assertThat(results[0]).isEqualTo("instances");
+  }
+
+  @Test
+  public void testAttributesToQidoPathArray_oneModalitySet() throws Exception {
+    Attributes attrs = new Attributes();
+    attrs.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+    attrs.setString(Tag.ModalitiesInStudy, VR.CS, "MG");
+
+    String[] results = AttributesUtil.attributesToQidoPathArray(attrs);
+
+    assertThat(results.length).isEqualTo(1);
+    assertThat(results[0]).isEqualTo(
+        "instances?" + TagUtils.toHexString(Tag.ModalitiesInStudy) + "=MG&");
+  }
+
+  @Test
+  public void testAttributesToQidoPathArray_manyModalitiesSet() throws Exception {
+    Attributes attrs = new Attributes();
+    attrs.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+    attrs.setString(Tag.ModalitiesInStudy, VR.CS, "MG", "CS", "CR");
+
+    String[] results = AttributesUtil.attributesToQidoPathArray(attrs);
+
+    assertThat(results.length).isEqualTo(3);
+    assertThat(results[0]).isEqualTo(
+        "instances?" + TagUtils.toHexString(Tag.ModalitiesInStudy) + "=MG&");
+    assertThat(results[1]).isEqualTo(
+        "instances?" + TagUtils.toHexString(Tag.ModalitiesInStudy) + "=CS&");
+    assertThat(results[2]).isEqualTo(
+        "instances?" + TagUtils.toHexString(Tag.ModalitiesInStudy) + "=CR&");
+  }
+
+  @Test
+  public void testJsonToAttributes_empty() throws Exception {
+    JSONObject jsonObj = new JSONObject();
+
+    Attributes attrs = AttributesUtil.jsonToAttributes(jsonObj);
+
+    assertThat(attrs).isEqualTo(new Attributes());
+  }
+
+  @Test
+  public void testJsonToAttributes_integerString() throws Exception {
+    // can be parsed as either string or int.
+    JSONObject jsonObj = new JSONObject("{\"" + TagUtils.toHexString(Tag.InstanceNumber)
+        + "\": {\"vr\": \"IS\",\"Value\": [1]}}");
```

Comment:
Some integer strings can exceed the max int32 (similar with the DS VR), so can we change the code to map these to strings?

Reply:
VRs that are both string and int are already parsed as strings. Added VR.DS unit test just in case.